### PR TITLE
KEYCLOAK-14192 Client Policy - Condition : Author of a client - User Role

### PR DESCRIPTION
**Pull request by tnorimat**
This PR is for [KEYCLOAK-14192 Client Policy - Condition : Author of a client - User Role](https://issues.redhat.com/browse/KEYCLOAK-14192) in [KEYCLOAK-13933 Client Policies](https://issues.redhat.com/browse/KEYCLOAK-13933), also is the part of the project [Client Policy Official Support](https://github.com/keycloak/kc-sig-fapi/projects) of [FAPI-SIG](https://github.com/keycloak/kc-sig-fapi) activity.

Generally speaking, the aim of this PR is to support policy conditions defined in [Client Policy design document](https://github.com/keycloak/keycloak-community/blob/master/design/client-policies.md#condition--which-client)

It behavior has been specified in [the JIRA ticket](https://issues.redhat.com/browse/KEYCLOAK-14192).

### CODE REVIEW

**Comment by tnorimat** [comment]
@mposolda This PR is about Client Policy Official Support I mentioned in [page 16 of FAPI-SIG 6th Meeting Agenda](https://github.com/keycloak/kc-sig-fapi/blob/master/FAPI-SIG/meetings/6th/presentations/FAPI-SIG_6th_MTG_agenda%20.pdf). Could you or your team review this PR?

File: services/src/main/java/org/keycloak/services/clientpolicy/condition/ClientUpdateSourceRolesCondition.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    private boolean isRolesMatched(UserModel user) {
+        if (user == null) return false;
+
+        Set<String> expectedRoles = instantiateRolesForMatching();
+        if (expectedRoles == null) return false;
+
+        // user.getRoleMappingsStream() never returns null according to {@link UserModel.getRoleMappingsStream}
+        Set<String> roles = user.getRoleMappingsStream().map(RoleModel::getName).collect(Collectors.toSet());
```
Comment (mposolda) [inline services/src/main/java/org/keycloak/services/clientpolicy/condition/ClientUpdateSourceRolesCondition.java:111]: It seems that this won't take composite roles into account? Perhaps just "user.hasRole" is better as it takes composite roles and also groups into account. For example case when user is member of group "g" and this group has role "foo" assigned, then the user is considered to be in role "foo" due the indirection through the group.

**Review by mposolda — CHANGES_REQUESTED** [review]
@tnorimat Thanks for the PR! I've added one inline comment.

File: services/src/main/java/org/keycloak/services/clientpolicy/condition/ClientUpdateSourceRolesCondition.java (hunk 1 above)
Comment (tnorimat) [reply services/src/main/java/org/keycloak/services/clientpolicy/condition/ClientUpdateSourceRolesCondition.java:111]: Yes. I understood the point not only treating the ordinal role but also the composit role. I'll use `user.hasRole` instaed of `user.getRoleMappingsStream`.

**Comment by tnorimat** [comment]
@mposolda I've fixed what pointed out and rebased.

**Review by mposolda — APPROVED** [review]
@tnorimat Thanks!

**Pull request merged**